lib/archethic/contracts.ex

```diff
@@ -165,6 +168,43 @@ defmodule Archethic.Contracts do
     end
   end
 
+  @doc """
+  Simulate the execution of the contract hold in prev_tx with the inputs of next_tx, at a certain date
+  """
+
+  @spec simulate_contract_execution(Transaction.t(), Transaction.t(), DateTime.t()) ::
+          %{valid: false, address: binary(), reason: binary()} | %{valid: true, address: binary()}
+  def simulate_contract_execution(
+        prev_tx = %Transaction{data: %TransactionData{code: code}},
+        next_tx = %Transaction{},
+        date = %DateTime{}
+      ) do
+    case Interpreter.parse(code) do
+      {:ok,
+       %Contract{
+         triggers: triggers,
+         conditions: %{inherit: transaction_conditions}
+       }} ->
+        constants = %{
+          "previous" => Constants.from_transaction(prev_tx),
+          "next" => Constants.from_transaction(next_tx)
+        }
+
+        with :ok <- validate_conditions(transaction_conditions, constants),
+             :ok <- validate_triggers(triggers, next_tx, date),
+             %Transaction{} <-
+               ActionInterpreter.execute(Map.fetch!(triggers, :transaction), constants) do
```

Comment:
`ActionInterpreter.execute` is not necessary returning a transaction. It might return nil and still be a valid contract (for example, when there is nothing to do)

Reply:
Hi Bastien, looking at the returns from `ActionInterpreter.execute` ( https://github.com/archethic-foundation/archethic-node/blob/develop/lib/archethic/contracts/interpreter/action.ex#L364 )  it seems a transaction is always returned ?

Reply:
Ah I see now I guess `next_transaction`, returned by `execute/2` might be `nil`. If this is asserted, I will update the specs for `execute/2`

Reply:
If `Nil` can on;y be returned when executing (  `ActionInterpreter.execute/2` ) with empty SC code. Then I think this might be ok because empty contracts are already checked at higher level in the code.

My issue is that it produce : 

``` %{"valid" => false, "address" => "00009e059e8171643b959284fe542909f3b32198b8fc25b3e50447589b84341c1d67", "reason" => "Transaction doesn't hold any contract"} ```

And I think we should rather consider empty contracts as valid instead ?

Reply:
Julien confirmed empty contracts should produce a valid result. I will change the code.